Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp

```diff
@@ -0,0 +1,181 @@
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "RenderSVGResourceMarker.h"
+
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+#include "Element.h"
+#include "ElementIterator.h"
+#include "FloatPoint.h"
+#include "Image.h"
+#include "ImageBuffer.h"
+#include "IntRect.h"
+#include "RenderLayer.h"
+#include "RenderLayerInlines.h"
+#include "RenderSVGModelObjectInlines.h"
+#include "RenderSVGResourceMarkerInlines.h"
+#include "SVGContainerLayout.h"
+#include "SVGLengthContext.h"
+#include "SVGRenderStyle.h"
+#include "SVGRenderingContext.h"
+#include "SVGResources.h"
+#include "SVGResourcesCache.h"
+#include <wtf/IsoMallocInlines.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGResourceMarker);
+
+RenderSVGResourceMarker::RenderSVGResourceMarker(SVGMarkerElement& element, RenderStyle&& style)
+    : RenderSVGResourceContainer(Type::SVGResourceMarker, element, WTFMove(style))
+{
+}
+
+RenderSVGResourceMarker::~RenderSVGResourceMarker() = default;
+
+void RenderSVGResourceMarker::invalidateMarker()
+{
+    // Repainting all clients is sufficient to handle invalidation.
+    repaintAllClients();
+}
+
+FloatRect RenderSVGResourceMarker::computeViewport() const
+{
+    auto& useMarkerElement = markerElement();
+    SVGLengthContext lengthContext(&useMarkerElement);
+    return { 0, 0, useMarkerElement.markerWidth().value(lengthContext), useMarkerElement.markerHeight().value(lengthContext) };
+}
+
+bool RenderSVGResourceMarker::updateLayoutSizeIfNeeded()
+{
+    auto previousViewportSize = viewportSize();
+    m_viewport = computeViewport();
+    return selfNeedsLayout() || previousViewportSize != viewportSize();
+}
+
+void RenderSVGResourceMarker::layout()
+{
+    // RenderSVGResourceContainer inherits from RenderSVGHiddenContainer which has a no-op layout() implementation.
+    // Markers are always painted indirectly, but their children need to be laid out such as they would be regular
+    // children of a SVG container element, thus skip RenderSVGHiddenContainer::layout() and use RenderSVGContainer::layout().
+    RenderSVGContainer::layout();
+}
+
+void RenderSVGResourceMarker::updateFromStyle()
+{
+    RenderSVGContainer::updateFromStyle();
+
+    if (SVGRenderSupport::isOverflowHidden(*this))
+        setHasNonVisibleOverflow();
```

Comment:
Just curious, why is this important? Which tests would fail otherwise? I saw the same code for patterns and it did not seem to make a difference.

Reply:
This handles "overflow: hidden" - only tested by WPT -- after this PR I plan to unskip svg/dynamic-updates + generate WPT baselines.